src/main/java/net/modfest/scatteredshards/client/screen/ShardCreatorGuiDescription.java

```diff
@@ -135,6 +139,74 @@ public static Identifier parseTexture(String path) {
 	private Identifier iconPath = null;
 
 
+	private <T> void updateComponents(StringReader reader) throws CommandSyntaxException {
+		ComponentChanges.Builder changesBuilder = ComponentChanges.builder();
+		Set<ComponentType<?>> known = new ReferenceArraySet<>();
+
+		// Begin of component list
+		reader.expect('[');
+		reader.skipWhitespace();
+
+		// Body of component list
+		while (reader.canRead() && reader.peek() != ']') {
+			boolean negation = false;
+
+			if (reader.peek() == '!') {
+				// Negate incoming block
+				reader.skip();
+				negation = true;
+			}
+
+			// Component Type
+			@SuppressWarnings("unchecked") // We could avoid this with a separate method for getting the values but eh
+			ComponentType<T> componentType = (ComponentType<T>) ItemStringReader.Reader.readComponentType(reader);
```

Comment:
shame we can't do better than an AW for a static call here, feels bad but hopefully this is all just exposed as a helper eventually